community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/ImportPanicIT.java

```diff
@@ -0,0 +1,168 @@
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.data;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.defaultFormatNodeFileHeader;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.defaultFormatRelationshipFileHeader;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.nodeData;
+import static org.neo4j.unsafe.impl.batchimport.input.csv.DataFactories.relationshipData;
+
+public class ImportPanicIT
+{
+    private static final int BUFFER_SIZE = 1000;
+
+    private final FileSystemAbstraction fs = new DefaultFileSystemAbstraction();
+    private final TestDirectory directory = testDirForTest( getClass() );
+    private final RandomRule random = new RandomRule();
+
+    @Rule
+    public final RuleChain rules = RuleChain.outerRule( random ).around( directory );
+
+    /**
+     * There was this problem where some steps and in particular parallel CSV input parsing that
+     * paniced would hang the import entirely.
+     */
+    @Test
+    public void shouldExitAndThrowExceptionOnPanic() throws Exception
+    {
+        // GIVEN
+        BatchImporter importer = new ParallelBatchImporter( directory.absolutePath(), fs, Configuration.DEFAULT,
+                NullLogService.getInstance(), ExecutionMonitors.invisible(), AdditionalInitialIds.EMPTY,
+                Config.empty() );
+        Iterable<DataFactory<InputNode>> nodeData =
+                nodeData( data( NO_NODE_DECORATOR, fileAsCharReadable( nodeCsvFileWithBrokenEntries() ) ) );
+        Input brokenCsvInput = new CsvInput(
+                nodeData, defaultFormatNodeFileHeader(),
+                relationshipData(), defaultFormatRelationshipFileHeader(),
+                IdType.ACTUAL,
+                csvConfigurationWithLowBufferSize(),
+                new BadCollector( new NullOutputStream(), 0, 0 ),
+                Runtime.getRuntime().availableProcessors() );
+
+        // WHEN
+        try
+        {
+            importer.doImport( brokenCsvInput );
+            fail( "Should have failed properly" );
+        }
+        catch ( InputException e )
+        {
+            // THEN
+            assertTrue( e.getCause() instanceof DataAfterQuoteException );
+            // and we managed to shut down properly
+        }
+    }
+
+    private org.neo4j.unsafe.impl.batchimport.input.csv.Configuration csvConfigurationWithLowBufferSize()
+    {
+        return new org.neo4j.unsafe.impl.batchimport.input.csv.Configuration.Overridden( COMMAS )
+        {
+            @Override
+            public int bufferSize()
+            {
+                return BUFFER_SIZE;
+            }
+        };
+    }
+
+    private Supplier<CharReadable> fileAsCharReadable( File file )
+    {
+        return () ->
+        {
+            try
+            {
+                return Readables.wrap( fs.openAsReader( file, Charset.defaultCharset() ) );
```

Comment:
Since only ASCII characters is used it doesn't matter. But sure using UTF-8 would have this test run with same charset on all environments. Is that good or bad?